### PR TITLE
Prevent attempting to reorder optional rows when unchanged

### PR DESCRIPTION
**Pull request by brancz**
In our system, a higher-level layer already ensures that all buffers are sorted, so we do a bunch of unnecessary work attempting to reorder a column whose order hasn't changed. This essentially copies the approach that the repeated buffers took.

### CODE REVIEW

File: column_buffer.go [hunk 1]
```diff
@@ -644,7 +646,21 @@ func (col *repeatedColumnBuffer) Values() ValueReader {
 	return &repeatedPageReader{page: col.Page().(*repeatedPage)}
 }
 
-func rowsHaveBeenReordered(rows []region) bool {
+func optionalRowsHaveBeenReordered(rows []int32) bool {
+	i := int32(0)
+	for _, row := range rows {
```
Comment (kevinburkesegment) [inline column_buffer.go:651]:
```go
for i := 0; i < len(rows); i++ {
    if i != rows[i]
}
```

?

File: column_buffer.go [hunk 2]
```diff
@@ -184,36 +184,38 @@ func (col *optionalColumnBuffer) Pages() Pages {
 }
 
 func (col *optionalColumnBuffer) Page() BufferedPage {
-	numNulls := countLevelsNotEqual(col.definitionLevels, col.maxDefinitionLevel)
-	numValues := len(col.rows) - numNulls
+	if optionalRowsHaveBeenReordered(col.rows) {
```
Comment (kevinburkesegment) [inline column_buffer.go:187]: Make the fast case more straightforward? I don't know

```go
if !optionalRowsHaveBeenReordered(col.rows) {
     	return newOptionalPage(col.base.Page(), col.maxDefinitionLevel, col.definitionLevels)
} 
...
```

Comment (brancz) [reply column_buffer.go:187]: Sure no problem, I was just replicating the same way the equivalent repeated buffer does it. Will do!

File: column_buffer.go (hunk 1 above)
Comment (brancz) [reply column_buffer.go:651]: That's not quite the same, we can only increase `i` if we actually see a non-null row.

Comment (brancz) [reply column_buffer.go:656]: I think this is not super obvious so I added some code comments that should clarify it for future readers.

**Comment by brancz** [comment]
I believe all comments on this PR have been addressed. Let me know if there is anything else!

**Review by achille-roussel — APPROVED** [review]
Looking good 👍

**Pull request merged**